src/peakWatcher.cpp

```diff
@@ -717,6 +730,172 @@ void report(int watcherIndex, int channel) {
 	outputMessage(s);
 }
 
+MediaTrack* getTrackFromGuidStr(ReaProject* project, const string& guidStr) {
+	if (guidStr == "MASTER") {
+		return GetMasterTrack(project);
+	}
+	GUID guid;
+	stringToGuid(guidStr.c_str(), &guid);
+	int trackCount = CountTracks(project);
+	for (int i = 0; i < trackCount; ++i) {
+		MediaTrack* track = GetTrack(project, i);
+		if (memcmp(&guid, GetTrackGUID(track), sizeof(GUID)) == 0) {
+			return track;
+		}
+	}
+	return nullptr;
+}
+
+string getTrackGuidStr(ReaProject* project, MediaTrack* track) {
+	if (track == GetMasterTrack(project)) {
+		// The master track doesn't have a persistent GUID.
+		return "MASTER";
+	}
+	char guid[64];
+	guidToString(GetTrackGUID(track), guid);
+	return guid;
+}
+
+/*
+ * Example config block:
+<OSARA_PEAKWATCHER
+  WATCHER TRACK {someGuid} TYPE 0 FOLLOW 1 LEVEL 0.0 HOLD 0 NOTIFY 0 0
+  WATCHER TRACKFX {someGuid} 5 TYPE 0 FOLLOW 0 LEVEL -5.0 HOLD -1 NOTIFY 1 1
+>
+ */
+
+const char CONFIG_HEADER[] = "<OSARA_PEAKWATCHER";
+const char CONFIG_FOOTER[] = ">";
+
+bool processExtensionLine(const char* line, ProjectStateContext* ctx,
+	bool isUndo, struct project_config_extension_t* reg
+) {
+	if (isUndo || strcmp(line, CONFIG_HEADER) != 0) {
+		return false;
+	}
+	stop();
+	ReaProject* project = GetCurrentProjectInLoadSave();
+	for (int w = 0; ; ++w) {
+		char data[500];
+		ctx->GetLine(data, sizeof(data));
+		if (strcmp(data, CONFIG_FOOTER) == 0) {
+			break;
+		}
+		if (w >= NUM_WATCHERS) {
+			continue;
+		}
+		istringstream input(data);
+		string word;
+		input >> word;
+		if (word != "WATCHER") {
+			continue;
+		}
+		Watcher& watcher = watchers[project][w];
+		watcher.reset();
+		input >> word;
+		if (word == "NONE") {
+			watcher.target = NoTarget();
+		} else if (word == "TRACK") {
+			input >> word;
+			if (MediaTrack* track = getTrackFromGuidStr(project, word)) {
+				watcher.target = track;
+			}
+		} else if (word == "TRACKFX") {
+			input >> word;
+			int fx;
+			input >> fx;
+			if (MediaTrack* track = getTrackFromGuidStr(project, word)) {
+				watcher.target = TrackFx(track, fx);
+			}
+		}
+		input >> word;
+		if (word == "TYPE") {
+			input >> watcher.levelType;
+		}
+		input >> word;
+		if (word == "FOLLOW") {
+			input >> word;
+			watcher.follow = word == "1";
+		}
+		input >> word;
+		if (word == "LEVEL") {
+			input >> watcher.notifyLevel;
+		}
+		input >> word;
+		if (word == "HOLD") {
+			input >> watcher.hold;
+		}
+		input >> word;
+		if (word == "NOTIFY") {
+			for (auto& channel : watcher.channels) {
+				input >> word;
+				channel.notify = word == "1";
+			}
+		}
+	}
+	if (isWatchingAnything()) {
```

Comment:
I was originally worried this was going to be a problem if the project being loaded isn't the active project. I'm not certain this can happen, but based on the fact that there is a separate GetCurrentProjectInLoadSave function, I'm guessing it can. Perhaps this happens if you load multiple projects at once; e.g. when starting REAPER with multiple tabs left open.

However, I'm now realising that this will probably be handled by the code which handles tab switching. Is that correct?

If so, we should still fix this to only assume we're watching something if GetCurrentProjectInLoadSave is actually the active project.

Reply:
Yes it is handled when switching tabs. I added a condition to only call start if loading the active project as you suggested.